File: app/src/main/java/lantian/nolitter/SettingsFragment.java
package lantian.nolitter;

import android.content.ComponentName;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.preference.PreferenceManager;
import android.support.v7.app.AlertDialog;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class SettingsFragment extends PreferenceFragment {
    public static SharedPreferences prefs;

    @Override
    @SuppressWarnings("deprecation")
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getPreferenceManager().setSharedPreferencesMode(Context.MODE_WORLD_READABLE);
        addPreferencesFromResource(R.xml.settings);
        prefs = PreferenceManager.getDefaultSharedPreferences(getActivity());
        getPreferenceScreen().removePreference(findPreference("hidden"));
        findPreference("banned_ui").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                ltChooseApp("banned", Constants.banned);
                return false;
            }
        });
        findPreference("forced_ui").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                ltChooseApp("forced", Constants.forced);
                return false;
            }
        });
        findPreference("chsdpath_ui").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                ltChooseApp("chsdpath", Constants.chsdpath);
                return false;
            }
        });
        findPreference("author").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://lantian.pub"));
                startActivity(browserIntent);
                return false;
            }
        });
        findPreference("source").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://github.com/xddxdd/lantian-nolitter"));
                startActivity(browserIntent);
                return false;
            }
        });
        findPreference("showicon").setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                Boolean isVisible = (Boolean) newValue;
                if (isVisible) {
                    getActivity().getPackageManager().setComponentEnabledSetting(new ComponentName(getActivity(), MainActivity.class),
                            PackageManager.COMPONENT_ENABLED_STATE_ENABLED,
                            PackageManager.DONT_KILL_APP);
                } else {
                    getActivity().getPackageManager().setComponentEnabledSetting(new ComponentName(getActivity(), MainActivity.class),
                            PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
                            PackageManager.DONT_KILL_APP);
                }
                return true;
            }
        });
    }

    public boolean ltChooseApp(final String key, final String defaults) {
        // Get app info
        final List<ApplicationInfo> appInfo;
        if (prefs.getBoolean("enable_system", false)) {
            appInfo = getActivity().getPackageManager().getInstalledApplications(PackageManager.GET_META_DATA);
        } else {
            appInfo = new ArrayList<>();
            for (ApplicationInfo appInfoSingle : getActivity().getPackageManager().getInstalledApplications(PackageManager.GET_META_DATA)) {
                if ((appInfoSingle.flags & ApplicationInfo.FLAG_SYSTEM) == 0) {
                    appInfo.add(appInfoSingle);
                }
            }
        }
        Collections.sort(appInfo, new ApplicationInfo.DisplayNameComparator(getActivity().getPackageManager()));

        // Get current settings
        ArrayList<String> banned = new ArrayList<>(Arrays.asList(prefs.getString(key, defaults).split(",")));

        // Form arrays
        final String[] appTitles = new String[appInfo.size()];
        boolean[] appCared = new boolean[appInfo.size()];
        for (int i = 0; i < appInfo.size(); i++) {
            appTitles[i] = appInfo.get(i).loadLabel(getActivity().getPackageManager()).toString();
            appCared[i] = banned.contains(appInfo.get(i).packageName);
        }

        // Make the dialog
        AlertDialog.Builder dialog = new AlertDialog.Builder(getActivity());
        dialog.setTitle(R.string.ui_chooseApp);
        dialog.setMultiChoiceItems(appTitles, appCared, new DialogInterface.OnMultiChoiceClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which, boolean isChecked) {
                ArrayList<String> banned = new ArrayList<>(Arrays.asList(prefs.getString(key, defaults).split(",")));
                if (isChecked) {
                    banned.add(appInfo.get(which).packageName);
                } else {
                    banned.remove(appInfo.get(which).packageName);
                }
                String newBanned = "";
                for (String bannedFragment : banned) {
                    if (!bannedFragment.trim().isEmpty()) newBanned += bannedFragment.trim() + ",";
                }
                prefs.edit().putString(key, newBanned).apply();
            }
        });
        dialog.setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.dismiss();
            }
        });
        dialog.show();
        return false;
    }
}

File: app/src/main/java/lantian/nolitter/SettingsFragment.kt
package lantian.nolitter

import android.content.*
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import java.util.*

class SettingsFragment : PreferenceFragmentCompat() {

    private var prefs: SharedPreferences? = null

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {

        setPreferencesFromResource(R.xml.settings, rootKey)
        prefs = context?.getSharedPreferences(BuildConfig.APPLICATION_ID + "_preferences", Context.MODE_WORLD_READABLE)

        findPreference<Preference>("forced_ui")?.setOnPreferenceClickListener {
            chooseApp("forced", Constants.forced)
            false
        }
        findPreference<Preference>("author")?.setOnPreferenceClickListener {
            val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse("https://lantian.pub"))
            startActivity(browserIntent)
            false
        }
        findPreference<Preference>("source")?.setOnPreferenceClickListener {
            val browserIntent =
                Intent(Intent.ACTION_VIEW, Uri.parse("https://github.com/xddxdd/lantian-nolitter"))
            startActivity(browserIntent)
            false
        }
        findPreference<Preference>("about_xinternalsd")?.setOnPreferenceClickListener {
            val browserIntent =
                Intent(Intent.ACTION_VIEW, Uri.parse("https://github.com/pylerSM/XInternalSD"))
            startActivity(browserIntent)
            false
        }
        findPreference<Preference>("show_icon")?.setOnPreferenceChangeListener { _, newValue ->
            requireActivity().packageManager.setComponentEnabledSetting(
                ComponentName(requireActivity(), MainActivity::class.java),
                if(newValue as Boolean) PackageManager.COMPONENT_ENABLED_STATE_ENABLED else PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
                PackageManager.DONT_KILL_APP
            )
            true
        }
    }

    private fun chooseApp(key: String?, defaults: String?): Boolean {
        // Get app info
        val appInfo: MutableList<ApplicationInfo>
        if (prefs!!.getBoolean("enable_system", false)) {
            appInfo = requireActivity().packageManager.getInstalledApplications(PackageManager.GET_META_DATA)
        } else {
            appInfo = ArrayList()
            for (appInfoSingle in requireActivity().packageManager.getInstalledApplications(PackageManager.GET_META_DATA)) {
                if (appInfoSingle.flags and ApplicationInfo.FLAG_SYSTEM == 0)
                    appInfo.add(appInfoSingle)
            }
        }
        Collections.sort(appInfo, ApplicationInfo.DisplayNameComparator(requireActivity().packageManager))

        // Get current settings
        val currentIgnored = ArrayList(listOf(*prefs!!.getString(key, defaults)!!.split(",".toRegex()).toTypedArray()))

        // Form arrays
        val appTitles = arrayOfNulls<String>(appInfo.size)
        val appCared = BooleanArray(appInfo.size)
        for (index in appInfo.indices) {
            appTitles[index] = appInfo[index].loadLabel(requireActivity().packageManager).toString()
            appCared[index] = currentIgnored.contains(appInfo[index].packageName)
        }

        // Make the dialog
        val alertDialog = AlertDialog.Builder(requireContext())
        alertDialog.setTitle(R.string.ui_chooseApp)
        alertDialog.setMultiChoiceItems(appTitles,appCared) { _, which: Int, isChecked: Boolean ->
            val ignored = ArrayList(listOf(*prefs!!.getString(key, defaults)!!.split(",".toRegex()).toTypedArray()))
            if (isChecked) {
                ignored.add(appInfo[which].packageName)
            } else {
                ignored.remove(appInfo[which].packageName)
            }
            var newIgnored = ""
            for (ignoredFragment in ignored) {
                if (ignoredFragment.trim { it <= ' ' }.isNotEmpty()) newIgnored += ignoredFragment.trim { it <= ' ' } + ","
            }
            prefs!!.edit().putString(key, newIgnored).apply()
        }
        alertDialog.setPositiveButton(android.R.string.ok) { dialog: DialogInterface, _ -> dialog.dismiss() }
        alertDialog.show()
        return false
    }
}